twarep/electrifly
# Weather data single flight bug

## Checklist

## What I Did
What files git you edit and why/for what purpose?

**scraper.py**: fixed problem where selecting months 10, 11, and 12 broke the scraper. Increased sleep time before choosing airport in weather data scrape.

**storage.py**: added two more cases to get all the weather data correctly. The original code only got weather data entries within the interval [flight start time, flight end time]. The two cases it missed were:

1. **If no weather data times occurred within the flight**: now the code looks at all the times before the interval on the same day, and selects the max one. Note that in one rare case, a flight occurred at a specific time and the closest weather reading was 5 hours earlier, which might be bad but this edge case is hard to take out without an additional if statement just for it.
2. **If there was weather data within the interval, but the first time didn't start directly at the flight start time:** This covers cases where a flight may have began at 11:04 AM, ended at 12:00 PM, but the only weather data corresponding was logged at 11:25 AM. This code just checks if this gap exists, and then gets the previous weather entry, which will always allow full coverage of the entire flight. 

## How I Did it?
Pandas work, use the [docs](https://pandas.pydata.org/docs/reference/index.html) to understand what I did here.

## How to Test?

Running this in the db: 
```sql 
SELECT id 
FROM flights 
WHERE id NOT IN 
(SELECT flight_id FROM flight_weather)
```
should return no results, as all flights should have a corresponding weather entry.
### Relevant Links
Any tutorials or other resources used to help with this task. You can also cite the documentation.
